tests/unit/s2n_client_hello_test.c

```diff
@@ -403,6 +403,110 @@ int main(int argc, char **argv)
+        uint8_t* sent_client_hello;
+        struct s2n_cert_chain_and_key *chain_and_key;
+
+        uint8_t client_extensions[] = {
+            /* Extension type TLS_EXTENSION_SERVER_NAME */
+            0x00, 0x00,
+            /* Extension size */
+            0x00, 0x08,
+            /* Server names len */
+            0x00, 0x06,
+            /* First server name type - host name */
+            0x00,
+            /* First server name len */
+            0x00, 0x03,
+            /* First server name, matches sent_server_name */
+            's', 'v', 'r',
+        };
+
+        int client_extensions_len = sizeof(client_extensions);
+        uint8_t client_hello_prefix[] = {
+            /* Protocol version TLS ??? */
+            0xFF, 0xFF,
+            /* Client random */
+            ZERO_TO_THIRTY_ONE,
+            /* SessionID len - 32 bytes */
+            0x20,
+            /* Session ID */
+            ZERO_TO_THIRTY_ONE,
+            /* Cipher suites len */
+            0x00, 0x02,
+            /* Cipher suite - TLS_RSA_WITH_AES_128_CBC_SHA256 */
+            0x00, 0x3C,
+            /* Compression methods len */
+            0x01,
+            /* Compression method - none */
+            0x00,
+            /* Extensions len */
+            (client_extensions_len >> 8) & 0xff, (client_extensions_len & 0xff),
+        };
+        int client_hello_prefix_len = sizeof(client_hello_prefix);
+        int sent_client_hello_len = client_hello_prefix_len + client_extensions_len;
+        uint8_t message_header[] = {
+            /* Handshake message type CLIENT HELLO */
+            0x01,
+            /* Body len */
+            (sent_client_hello_len >> 16) & 0xff, (sent_client_hello_len >> 8) & 0xff, (sent_client_hello_len & 0xff),
+        };
+        int message_len = sizeof(message_header) + sent_client_hello_len;
+        uint8_t record_header[] = {
+            /* Record type HANDSHAKE */
+            0x16,
+            /* Protocol version TLS 1.2 */
+            0x03, 0x03,
+            /* Message len */
+            (message_len >> 8) & 0xff, (message_len & 0xff),
+        };
+
+        EXPECT_NOT_NULL(sent_client_hello = malloc(sent_client_hello_len));
+        memcpy_check(sent_client_hello, client_hello_prefix, client_hello_prefix_len);
+        memcpy_check(sent_client_hello + client_hello_prefix_len, client_extensions, client_extensions_len);
+
+        /* Create nonblocking pipes */
+        EXPECT_SUCCESS(pipe(server_to_client));
+        EXPECT_SUCCESS(pipe(client_to_server));
+        for (int i = 0; i < 2; i++) {
+            EXPECT_NOT_EQUAL(fcntl(server_to_client[i], F_SETFL, fcntl(server_to_client[i], F_GETFL) | O_NONBLOCK), -1);
+            EXPECT_NOT_EQUAL(fcntl(client_to_server[i], F_SETFL, fcntl(client_to_server[i], F_GETFL) | O_NONBLOCK), -1);
+        }
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        server_conn->server_protocol_version = S2N_TLS12;
```

Comment:
It doesn't need to be set explicitly for this test to work, right?

Reply:
Good point. This is testing version negotiation and we initialize these values when we create the connection objects. I'll purge this line and add a check for `actual_protocol_version` after we call `s2n_negotiate`